extensions/emmet/src/abbreviationActions.ts

```diff
@@ -37,45 +43,166 @@ export function wrapWithAbbreviation(args: any) {
+		const matches = firstLineOfSelection.match(/^(\s*)/);
+		const preceedingWhiteSpace = matches ? matches[1].length : 0;
+
+		rangeToReplace = new vscode.Range(rangeToReplace.start.line, rangeToReplace.start.character + preceedingWhiteSpace, rangeToReplace.end.line, rangeToReplace.end.character);
+		let textToReplace = editor.document.getText(rangeToReplace);
+		rangesToReplace.push({ currentRange: rangeToReplace, originalRange: rangeToReplace, content: textToReplace });
+	});
+
+	let abbreviationPromise;
+	let currentValue = '';
+
+	function inputChanged(value: string): string {
+		if (value !== currentValue) {
+			currentValue = value;
+			makeChanges(value, previewMade, false).then((out) => {
+				if (typeof out === 'boolean') {
+					previewMade = out;
+				}
+			});
+		}
+		return '';
+	}
+
+	abbreviationPromise = (args && args['abbreviation']) ? Promise.resolve(args['abbreviation']) : vscode.window.showInputBox({ prompt: 'Enter Abbreviation', validateInput: inputChanged });
 	const helper = getEmmetHelper();
 
-	return abbreviationPromise.then(inputAbbreviation => {
-		if (!inputAbbreviation || !inputAbbreviation.trim() || !helper.isAbbreviationValid(syntax, inputAbbreviation)) { return false; }
+	function makeChanges(inputAbbreviation: string | undefined, previewMade?: boolean, definitive?: boolean): Thenable<any> {
+		if (!inputAbbreviation || !inputAbbreviation.trim() || !helper.isAbbreviationValid(syntax, inputAbbreviation)) {
+			let returnPromise: Thenable<any> = Promise.resolve();
+			if (previewMade) {
+				returnPromise = revertPreview(editor, rangesToReplace).then(() => { return false; });
+			}
+			return returnPromise;
+		}
 
 		let extractedResults = helper.extractAbbreviationFromText(inputAbbreviation);
 		if (!extractedResults) {
-			return false;
+			return Promise.resolve(previewMade);
 		}
 		let { abbreviation, filter } = extractedResults;
 
-		let expandAbbrList: ExpandAbbreviationInput[] = [];
+		expandAbbrList = [];
 
-		editor.selections.forEach(selection => {
-			let rangeToReplace: vscode.Range = selection.isReversed ? new vscode.Range(selection.active, selection.anchor) : selection;
-			if (rangeToReplace.isEmpty) {
-				let { active } = selection;
-				let currentNode = getNode(rootNode, active, true);
-				if (currentNode && (currentNode.start.line === active.line || currentNode.end.line === active.line)) {
-					rangeToReplace = new vscode.Range(currentNode.start, currentNode.end);
-				} else {
-					rangeToReplace = new vscode.Range(rangeToReplace.start.line, 0, rangeToReplace.start.line, editor.document.lineAt(rangeToReplace.start.line).text.length);
-				}
+		// we need to apply the previewchanges and get the new ranges
+		let revertPromise: Thenable<any> = Promise.resolve();
+		if (definitive) {
+			if (previewMade) {
+				revertPromise = revertPreview(editor, rangesToReplace);
 			}
+			rangesToReplace.forEach(rangesAndContent => {
```

Comment:
The `currentRange` of each `rangesToReplace` array gets updated inside `revertPreview`. Since the `originalRange` doesnt change, it is safe to loop through and use it here before `revertPromise` is completed. But for the sake of readability, it is better to do this after `revertPromise is completed.